### PR TITLE
fix mypy typing for model summary

### PR DESCRIPTION
**Pull request by awaelchli**
## What does this PR do?

Part of https://github.com/PyTorchLightning/pytorch-lightning/issues/7037
Follow up to #9344 

## Before submitting
- [x] Was this discussed/approved via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your PR does only one thing, instead of bundling different changes together?
- [x] Did you make sure to update the documentation with your changes? (if necessary)
- [x] Did you write any new necessary tests? (not for typos and docs)
- [x] Did you verify new and existing tests pass locally with your changes?
- [x] Did you update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)? (not for typos, docs, test updates, or internal minor changes/refactorings)

## PR review
Anyone in the community is free to review the PR once the tests have passed.
Before you start reviewing make sure you have read [Review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

 - [ ] Is this pull request ready for review? (if not, please submit in draft mode)
 - [x] Check that all items from **Before submitting** are resolved
 - [x] Make sure the title is self-explanatory and the description concisely explains the PR
 - [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?
I made sure I had fun coding 🙃

### CODE REVIEW

File: pytorch_lightning/utilities/model_summary.py [hunk 1]
```diff
@@ -310,9 +312,9 @@ def _get_summary_data(self):
             arrays.append(["In sizes", self.in_sizes])
             arrays.append(["Out sizes", self.out_sizes])
 
-        return arrays
+        return arrays  # type: ignore
```
Comment (awaelchli) [inline pytorch_lightning/utilities/model_summary.py:315]: `pytorch_lightning/utilities/model_summary.py:315: error: Incompatible return value type (got "List[List[Sequence[str]]]", expected "List[List[Union[str, List[str]]]]")`

why???

Comment (carmocca) [reply pytorch_lightning/utilities/model_summary.py:315]: I'm not exactly sure about why it converts it to `Sequence`, however, the return annotation doesn't seem right to me.

`List[Union[str, List[str]]]` suggests that it can be of any length and that `str` and `List[str]` can be in any order, however, the actual format represents just two elements where the first is a `str` and the second a `List[str]`.

For that reason, I'd annotate this return type as `List[Tuple[str, List[str]]]` and update the code to this stricter requirement:

```python
        arrays = [
            (" ", [str(i) for i in range(len(self._layer_summary))]),
            ("Name", self.layer_names),
            ("Type", self.layer_types),
            ("Params", [get_human_readable_count(n) for n in self.param_nums]),
        ]
        if self._model.example_input_array is not None:
            arrays.append(("In sizes", self.in_sizes))
            arrays.append(("Out sizes", self.out_sizes))
        return arrays
```

note: list comprehensions are more pythonic than `list(map(str` :D

Comment (awaelchli) [reply pytorch_lightning/utilities/model_summary.py]: great, thanks that works. I won't change the rest of the code, I want this PR to be just about typing.

File: pytorch_lightning/callbacks/model_summary.py [hunk 2]
```diff
@@ -46,11 +46,11 @@ class ModelSummary(Callback):
         >>> trainer = Trainer(callbacks=[ModelSummary(max_depth=1)])
     """
 
-    def __init__(self, max_depth: Optional[int] = 1):
+    def __init__(self, max_depth: int = 1) -> None:
         self._max_depth: int = max_depth
 
     def on_pretrain_routine_start(self, trainer: "pl.Trainer", pl_module: "pl.LightningModule") -> None:
-        if trainer.is_global_zero and self._max_depth is not None and not trainer.testing:
+        if trainer.is_global_zero and self._max_depth and not trainer.testing:
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_summary.py:53]: this can be a separate PR, but creating the summary should not be gated by `is_global_zero` here. This will break for DeepSpeed/FSDP or any model with parameters on different devices in the forward pass if the LightningModule sets an `example_input_array`. `forward` needs to act like a collective, so it must run on all ranks at the same time. it's one of the points listed in https://github.com/PyTorchLightning/pytorch-lightning/issues/9043

It is safer to generate the summary data unconditionally on all ranks and then let the logging of the summarization be determined by the particular rank id

File: pytorch_lightning/callbacks/model_summary.py [hunk 3]
```diff
@@ -62,11 +62,10 @@ def on_pretrain_routine_start(self, trainer: "pl.Trainer", pl_module: "pl.Lightn
 
     @staticmethod
     def summarize(
-        summary_data: List[List[Union[str, List[str]]]],
+        summary_data: List[Tuple[str, List[str]]],
```
Comment (ananthsub) [inline pytorch_lightning/callbacks/model_summary.py:65]: @kaushikb11 - if this set of arguments is going to grow, should we use a dataclass instead? that way the `summarize` signature doesn't change, and we can still add more arguments over time w/o breaking backwards compatibility or making everything optional.

File: pytorch_lightning/callbacks/model_summary.py [hunk 4]
```diff
@@ -46,11 +46,11 @@ class ModelSummary(Callback):
         >>> trainer = Trainer(callbacks=[ModelSummary(max_depth=1)])
     """
 
-    def __init__(self, max_depth: Optional[int] = 1):
+    def __init__(self, max_depth: int = 1) -> None:
         self._max_depth: int = max_depth
```
Comment (awaelchli) [inline pytorch_lightning/callbacks/model_summary.py:50]: reviewer: here in the new callback we don't need to type it None, because it does not mix with the deprecated "mode" argument that the `ModelSummary` has.

File: pytorch_lightning/utilities/model_summary.py [hunk 5]
```diff
@@ -85,19 +85,20 @@ def _register_hook(self) -> Optional[RemovableHandle]:
             A handle for the installed hook, or ``None`` if registering the hook is not possible.
         """
 
-        def hook(module, inp, out):
+        def hook(_: nn.Module, inp: Any, out: Any) -> None:
             if len(inp) == 1:
                 inp = inp[0]
             self._in_size = parse_batch_shape(inp)
             self._out_size = parse_batch_shape(out)
+            assert self._hook_handle is not None
```
Comment (awaelchli) [inline pytorch_lightning/utilities/model_summary.py:93]: reviewer: the outer function here registers the hook, so the handle must not be None. and if it was, it would mean the hook would not run. so this assert is good.

File: pytorch_lightning/utilities/model_summary.py [hunk 6]
```diff
@@ -207,21 +208,22 @@ def __init__(self, model, mode: Optional[str] = None, max_depth: Optional[int] =
...
         self._layer_summary = self.summarize()
         # 1 byte -> 8 bits
-        # TODO: how do we compute precisin_megabytes in case of mixed precision?
+        # TODO: how do we compute precision_megabytes in case of mixed precision?
         precision = self._model.precision if isinstance(self._model.precision, int) else 32
         self._precision_megabytes = (precision / 8.0) * 1e-6
 
     @property
     def named_modules(self) -> List[Tuple[str, nn.Module]]:
+        mods: List[Tuple[str, nn.Module]]
```
Comment (awaelchli) [inline pytorch_lightning/utilities/model_summary.py:217]: mypy doesn't get it, so I had to put it here

File: pytorch_lightning/utilities/model_summary.py [hunk 7]
```diff
@@ -450,14 +457,16 @@ def summarize(
 
     # temporary mapping from mode to max_depth
     if max_depth is None:
-        if mode in ModelSummaryMode.supported_types():
+        if mode is None:
+            model_summary = ModelSummary(lightning_module, max_depth=1)
+        elif mode in ModelSummaryMode.supported_types():
```
Comment (awaelchli) [inline pytorch_lightning/utilities/model_summary.py:462]: reviewer: had to change the if-else this structure here. Otherwise we could have ended up with a reference to the variable `model_summary` before assignment

File: pytorch_lightning/utilities/model_summary.py [hunk 8]
```diff
@@ -430,8 +437,8 @@ def _is_lazy_weight_tensor(p: Tensor) -> bool:
 
 
 def summarize(
-    lightning_module: "pl.LightningModule", mode: Optional[str] = "top", max_depth: Optional[int] = None
-) -> Optional[ModelSummary]:
+    lightning_module: "pl.LightningModule", mode: Optional[str] = None, max_depth: Optional[int] = None
+) -> ModelSummary:
```
Comment (awaelchli) [inline pytorch_lightning/utilities/model_summary.py:441]: this should have been the correct way in the original PR. mode is deprecated, we want to set it to None and then detect if someone sets it so we can warn appropriately.

File: pytorch_lightning/callbacks/model_summary.py (hunk 2 above)
Comment (awaelchli) [reply pytorch_lightning/callbacks/model_summary.py:53]: yes I agree. only guard the logging instruction

Comment (awaelchli) [reply pytorch_lightning/callbacks/model_summary.py:53]: also the condition  not trainer.testing here is always true, the hook here only runs for training :)

**Review by ananthsub — APPROVED** [review]

**Review by carmocca — APPROVED** [review]

File: pytorch_lightning/callbacks/model_summary.py (hunk 3 above)
Comment (kaushikb11) [reply pytorch_lightning/callbacks/model_summary.py:65]: @ananthsub The idea makes sense but I am not sure what additional arguments we will be planning to add in the future.

**Review by kaushikb11 — APPROVED** [review]
